RFCs/FS-1090-Generic-struct-type-whose-fields-are-all-unmanaged-types-is-unmanaged.md

```diff
@@ -0,0 +1,130 @@
+# F# RFC FS-1088 - (Generic struct type whose fields are all unmanaged types is unmanaged)
+
+The design suggestion [Generic struct type whose fields are all unmanaged types is unmanaged](https://github.com/fsharp/fslang-suggestions/issues/692) has been marked "approved in principle".
+This RFC covers the detailed proposal for this suggestion.
+
+* [x] Approved in principle
+* [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/692)
+* [ ] Implementation: [In progress](https://github.com/dotnet/fsharp/pull/6064)
+
+# Summary
+
+[summary]: #summary
+
+Allow generic structs to be `unmanaged` at construction if all fields are `unmanaged` types.
+
+# Motivation
+
+[motivation]: #motivation
+
+* Feature parity with C#.
+* Improve interoperability with C#.
+
+# Detailed design
+
+[design]: #detailed-design
+
+Currently, it is not possible to have an `unmanaged` constraint working on generic struct types.
+
+Consider following code:
+
+```fsharp
+[<Struct>]
+type Test<'T when 'T: unmanaged> =
+        val element: 'T
+
+let works = Test<int>()
+let error = Test<Test<int>>()
+
+```
+
+Prior to this RFC, the example above will fail to compile with:
+
+```less
+let error = Test<Test<int>>()
+  ------------^^^^^^^^^^^^^^^
+
+stdin(6,13): error FS0001: A generic construct requires that the type 'Test<int>' is an unmanaged type
+```
+
+This proposal aims to eliminate this restriction by treating generic struct type `unmanaged` if all its fields are `unmanaged`.
+
```

Comment:
Here we'll need a few things:

* What the definition of an unmanaged type is
* How the existing [`umanaged`](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/generics/constraints) constraint is to be adjusted (if at all)
* All types that are considered unmanaged (e.g., the numeric primitives)
* Example of custom struct that is unmanaged and one that is not
* How do we consider struct records? Allowed, subject to specific rules, disallowed, etc.
* How do we consider struct anonymous records? Allowed, subject to specific rules, disallowed, etc.
* How do we consider struct single-case DUs? Allowed, subject to specific rules, disallowed, etc.
* How do we consider struct multi-case DUs? Allowed, subject to specific rules, disallowed, etc.
* How do we deal with the `modreq` on existing APIs marked with them for this constraint?
* Do we also emit an `modreq`?
* C# has their [own metadata representation ](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.3/blittable#metadata-representation)since the F# constraint put them in a little design hole. What do we do about that?

(Note on `modreq` - we don't really use it correctly. @dsyme and @TIHan can fill you in on that bag of fun)

Reply:
I've updated examples, supported types, posted a description of the unmanaged types, etc.
I will need some time to figure out `modreq` part though.